File: lib/rayz/csg.rb
require_relative "shape"

module Rayz
  class CSG < Shape
    attr_reader :operation, :left, :right

    def initialize(operation:, left:, right:)
      super()
      @operation = operation
      @left = left
      @right = right
      # Set parent relationships
      @left.parent = self
      @right.parent = self
    end

    def local_intersect(local_ray)
      # Intersect ray with both children
      left_xs = @left.intersect(local_ray)
      right_xs = @right.intersect(local_ray)

      # Combine and sort intersections
      xs = (left_xs + right_xs).sort_by(&:t)

      # Filter based on CSG operation
      filter_intersections(xs)
    end

    def local_normal_at(local_point)
      # CSG objects have no surface of their own
      raise "CSG shapes have no surface and cannot have normals computed"
    end

    def includes?(shape)
      @left.includes?(shape) || @right.includes?(shape)
    end

    private

    def filter_intersections(xs)
      # Track whether we're inside each child shape
      inl = false
      inr = false
      result = []

      xs.each do |i|
        # Determine if intersection is on left or right child
        lhit = @left.includes?(i.object)

        # Check if this intersection is allowed
        if intersection_allowed?(lhit, inl, inr)
          result << i
        end

        # Toggle inside/outside state for the hit shape
        if lhit
          inl = !inl
        else
          inr = !inr
        end
      end

      result
    end

    def intersection_allowed?(lhit, inl, inr)
      case @operation
      when "union"
        (lhit && !inr) || (!lhit && !inl)
      when "intersection"
        (lhit && inr) || (!lhit && inl)
      when "difference"
        (lhit && !inr) || (!lhit && inl)
      else
        false
      end
    end
  end

  # Module-level helper function for testing
  def self.intersection_allowed(op, lhit, inl, inr)
    case op
    when "union"
      (lhit && !inr) || (!lhit && !inl)
    when "intersection"
      (lhit && inr) || (!lhit && inl)
    when "difference"
      (lhit && !inr) || (!lhit && inl)
    else
      false
    end
  end

  # Module-level helper function for testing
  def self.filter_intersections(csg, xs)
    csg.send(:filter_intersections, xs)
  end

  # Helper function to create CSG objects
  def self.csg(operation, left, right)
    CSG.new(operation: operation, left: left, right: right)
  end
end

File: python/rayz/csg.py
from __future__ import annotations

from rayz.shape import Shape
from rayz.tuple import Vector


class CSG(Shape):
    def __init__(self, operation: str, left: Shape, right: Shape) -> None:
        super().__init__()
        self.operation = operation
        self.left = left
        self.right = right
        self._bounds_cache = None
        left.parent = self
        right.parent = self

    def invalidate_bounds_cache(self) -> None:
        self._bounds_cache = None
        super().invalidate_bounds_cache()

    def includes(self, shape) -> bool:
        return self.left.includes(shape) or self.right.includes(shape)

    def local_intersect(self, ray) -> list:
        if not self.bounds().intersects(ray):
            return []
        left_xs = self.left.intersect(ray)
        right_xs = self.right.intersect(ray)
        xs = sorted(left_xs + right_xs, key=lambda i: i.t)
        return filter_intersections(self, xs)

    def local_normal_at(self, point, hit=None) -> Vector:
        raise RuntimeError("CSG shapes have no surface normal")

    def bounds(self):
        if self._bounds_cache is None:
            l_bounds = self.left.bounds().transform(self.left.transform)
            r_bounds = self.right.bounds().transform(self.right.transform)
            self._bounds_cache = l_bounds.merge(r_bounds)
        return self._bounds_cache


def intersection_allowed(op: str, lhit: bool, inl: bool, inr: bool) -> bool:
    if op == "union":
        return (lhit and not inr) or (not lhit and not inl)
    if op == "intersection":
        return (lhit and inr) or (not lhit and inl)
    if op == "difference":
        return (lhit and not inr) or (not lhit and inl)
    return False


def filter_intersections(csg: CSG, xs: list) -> list:
    inl, inr = False, False
    result = []
    for i in xs:
        lhit = csg.left.includes(i.object)
        if intersection_allowed(csg.operation, lhit, inl, inr):
            result.append(i)
        if lhit:
            inl = not inl
        else:
            inr = not inr
    return result
